cluster-up/cluster/kind-k8s-sriov-1.17.0/config_sriov.sh

```diff
@@ -99,93 +83,54 @@ function wait_k8s_object {
 
   local -r action="_kubectl get $object_type $name $namespace -o custom-columns=NAME:.metadata.name --no-headers"
 
-  retry "$tries" "$wait_time" "$action" "$wait_message" && return 0
-  echo $error_message && return  1
-}
-
-
-function is_taint_absence {
-  local -r taint=$1
-
-  result=$(_kubectl get nodes -o jsonpath="{.items[*].spec.taints[?(@.effect == \"$taint\")].effect}" || echo error)
-  if [[ -z $result ]]; then
-    echo "not-present"
+  if ! retry "$tries" "$wait_time" "$action" "$wait_message";then 
+    echo $error_message
+    return  1
   fi
-}
-
-function wait_for_taint_absence {
-  local -r taint=$1
 
-  local -r tries=60
-  local -r wait_time=5
-
-  local -r wait_message="Waiting for $taint taint absence"
-  local -r error_message="Taint $taint $name did not removed"
-  local -r action="is_taint_absence $taint"
-
-  retry "$tries" "$wait_time" "$action" "$wait_message" && return 0
-  echo $error_message && return 1
-}
-
-function wait_for_taint {
-  local -r taint=$1
-
-  local -r tries=60
-  local -r wait_time=5
-
-  local -r wait_message="Waiting for $taint taint to present"
-  local -r error_message="Taint $taint $name did not present"
-  local -r action="_kubectl get nodes -o custom-columns=taints:.spec.taints[*].effect --no-headers | grep -i $taint"
-
-  retry "$tries" "$wait_time" "$action" "$wait_message" && return 0
-  echo $error_message && return 1
+  return 0
 }
 
 # not using kubectl wait since with the sriov operator the pods get restarted a couple of times and this is
 # more reliable
 function wait_pods_ready {
-    while [ -n "$(_kubectl get pods --all-namespaces -o'custom-columns=status:status.containerStatuses[*].ready,metadata:metadata.name' --no-headers | grep false)" ]; do
-        echo "Waiting for all pods to become ready ..."
-        _kubectl get pods --all-namespaces -o'custom-columns=status:status.containerStatuses[*].ready,metadata:metadata.name' --no-headers
-        sleep 10
-    done
-}
-
-function wait_pods_condition_by_label {
-  local -r tries=10
-  local -r wait_time=1
-  local -r namespace=$1
-  local -r label=$2
-  local -r condition=$3
+  local -r tries=30
+  local -r wait_time=10
 
-  local -r wait_message="waiting for pod label $label at namespace $namespace to meet condition $condition"
-  local -r error_message="pod label $label at namespace $namespace did not meet condition $condition"
-  local -r action="_kubectl wait pods -n $namespace -l $label --for condition=$condition --timeout 30s >/dev/null && echo succeed"
+  local -r wait_message="Waiting for all pods to become ready.."
+  local -r error_message="Not all pods were ready after $(($tries*$wait_time)) seconds"
+  local -r get_pods_status="cluster-up/kubectl.sh get pods --all-namespaces -ocustom-columns=status:status.conditions[?(@.type==\"Ready\")].status,metadata:metadata.name --no-headers"
+  local -r action='[[ -z $($get_pods_status | grep -i false) ]] && echo all-pods-ready'
```

Comment:
In case `$($get_pods_status | grep -i false)` fails, the condition will be still true, since STDOUT might be empty. 
The right method is the look also on the return code, and only if it succeed, to check its stdout + grep please

Reply:
When `kubectl` fails on `get` operations it throws an error and a message to STDOUT
We shouldn't add another operation to that action

Reply:
It throw it to STDERR, and the grep doesn't catch this.

Reply:
The output from your example is 
```
$ [[ -z $(cluster-up/kubectl.sh get pods | grep configuration) ]] && echo bla
No resources found in default namespace.
bla
```
in fact if you try to get pods from a namespace that has no pods you would expect to get an error 
but acctially the return code will be '0' and the output will be
```
$ kubectl get po
No resources found in default namespace.
$ echo $?
0
```

Reply:
Please use my command, with real error, without KUBECONFIG and without cluster.
and see the bottom line, you can't relay on a failed command and then `grep -i words`

Reply:
This is also correct if you try to get pod from non existing namespace
```
$ kubectl get po -n blabla
No resources found in blabla namespace.
$ echo $?
0
```

Reply:
I don't want the command to success, i shown an example of a failed command, where the grep doesn't catch STDERR.
and asking please to focus on the bottom line only,
the PR as its now relays on a command that might fail, without taking this possible situation into account, but proceeding without even retry as it should.

Reply:
I am not aware to a scenario where we get to config_sriov.sh script to `wait_pods_ready` w/o a cluster or a kubeconfig
if we do, `wait_pods_ready` will return an error.
```
# kubectl does not exists
$ wait_pods_ready 
bash: ./kubectl: No such file or directory

# no kubeconfig
$ wait_pods_ready 
W1202 11:43:04.521749   11169 loader.go:223] Config not found: asdThe connection to the server localhost:8080 was refused - did you specify the right host or port?

# not valid kubeconfig
$ wait_pods_ready 
error: error loading config file "/root/go/src/kubevirt.io/kubevirt/_ci-configs/kind-k8s-sriov-1.17.0/.kubeconfig": v1.Config.Contexts: []v1.NamedContext: Clusters: []v1.NamedCluster: v1.NamedCluster.Name: Cluster: v1.Cluster.Server: CertificateAuthorityData: decode base64: illegal base64 data at input byte 1371, error found in #10 byte of ...|0tLQoasd=","server":|..., bigger context ...|p4R2pmST0K

# no cluster
$  wait_pods_ready 
The connection to the server 127.0.0.1:32767 was refused - did you specify the right host or port?
```
Those changes already been reviews and approved, CI is green and the cases you mention tested manually on local setup.

If you insist this should be change lets hear other reviewers thoughts.

Reply:
> I am not aware to a scenario where we get to config_sriov.sh script to `wait_pods_ready` w/o a cluster or a kubeconfig
> if we do, `wait_pods_ready` will return an error.

It was just for the example that a failed command will pass silently on the refereed line.

> Those changes already been reviews and approved, CI is green and the cases you mention already been tested manually on a local setup.
> If you insist this should be change lets hear other reviewers thoughts.

I totally insist, its a bug.

Reply:
I will elaborate on how `wait_pods_ready` works 
`wait_pods_ready` make use of `retry` and as you already know it will stop and return 0 once the action we pass does not write to STDOUT
We cant  relay on the return code when working with `kubectl` sometimes it doesn't return error code as you would expect it to return (discussed on other threads)

`[[ -z $($get_pods_status | grep -i false) ]] && echo all-pods-ready` passed as the `action` to `retry`
if there is a pods that is not ready, it will print `all-pods-ready` to STDOUT and `retry` will continue.
It will stop once all pods are ready, no 'false' found by grep, print 'all-pods-ready' to STDOUT and `retry` stop.

The way I see it `wait_pods_ready` acts as expected and there is no actual problem here
in case there is no kubeconfig or a cluster, cluster-up would stop before we get to this script.

@qinqon what you think?

Reply:
Notice that git diff shows that `wait_pods_condition_by_label` is changed but it actually removed and `wait_pods_ready`
is changed.
I was talking about `wait_pods_ready`

This PR intent is to simplify the way we indicate that the cluster is ready after sriov-operator installation, see PR description
long story short we relay on checking that all pods are ready and we cant just get into infinity loop.

Reply:
I totally in the loop and understand what PR does,
i am just saying that 
`  local -r action="_kubectl wait pods -n $namespace -l $label --for condition=$condition --timeout 30s >/dev/null && echo succeed"`
is exactly what will solve the all you need in this thread.

instead of those
```
  local -r get_pods_status="cluster-up/kubectl.sh get pods --all-namespaces -ocustom-columns=status:status.conditions[?(@.type==\"Ready\")].status,metadata:metadata.name --no-headers"
  local -r action='[[ -z $($get_pods_status | grep -i false) ]] && echo all-pods-ready'
```
and the problems and the refactor of retry which has problems imo (on the other thread).

Reply:
Now that you put it like that I understand
I also tested `wait_pods_ready` and saw it dosnt propagates the errors, good catch!

I like what you suggest, I tested it and it but seems that using 
`kubectl wait pods -A --all --for condition=Ready --timeout 0s` takes about 2~3 seconds to finish while using `kubectl pods get` takes less then a second, it means that it will take for `wait_pods_ready` extra 45\~75 seconds to finish.

I started to investigate how to do it w/o using `kubectl wait .. --for` and it seems that there is no simple kubectl command to get all pods that are not ready, where not ready means at least one container is not ready.
I have managed to make use of `kubectl get pods .. ` combined with `grep` by extracting it to another function and handle the errors there, so no extra time added and now `wait_pods_ready` propagates the errors properly.
it was tested with and w/o kubeconfig, when all pods ready and when there is a pod that not all of its containers are ready.
I will push it with next commits.

Reply:
I don't understand why it would take that much, and i prefer to use please the most keep it simple approach.
we see here that using the `wait pods` can solve at once the two threads,
this one and the one that changed the retry function, and has some problems imo (see the thread of that please).

Reply:
I tested it and saw it takes more time, actually the time related to the number of pods you have in the cluster.